### PR TITLE
Add mpas-framework support for several other attributes

### PR DESCRIPTION
**Pull request by jonbob**
Adds parsing support to the mpas-framework for the following attributes:
* bounds
* cell_measures
* cell_method
* coordinates
* standard_name

[BFB]

### CODE REVIEW

File: components/mpas-framework/src/tools/registry/gen_inc.c [hunk 1]
```diff
@@ -1618,6 +1731,86 @@ int parse_var(FILE *fd, ezxml_t registry, ezxml_t superStruct, ezxml_t currentVa
...
+			token = strsep(&string, "'");
+			sprintf(temp_str, "%s", token);
+
+			while ( ( token = strsep(&string, "'") ) != NULL ) {
+				sprintf(temp_str, "%s''%s", temp_str, token);
+			}
+
+			free(tofree);
+
+			fortprintf(fd, "         call mpas_add_att(%s %% attLists(const_index) %% attList, 'bounds', '%s')\n", pointer_name_arr, temp_str);
```
Comment (xylar) [inline components/mpas-framework/src/tools/registry/gen_inc.c:1747]: Here and for the remaining attributes except `standard_name`, I think this is over-indented compared with adjacent code.  Here's what I'm seeing:
```
      call mpas_add_att(r1aPtr(1) % attLists(1) % attList, 'units', 'm')
         call mpas_add_att(r1aPtr(1) % attLists(const_index) % attList, 'bounds', 'lonVertex')
         call mpas_add_att(r1aPtr(1) % attLists(const_index) % attList, 'cell_measures', 'area: areaCell')
         call mpas_add_att(r1aPtr(1) % attLists(const_index) % attList, 'cell_method', 'area: mean where sea')
         call mpas_add_att(r1aPtr(1) % attLists(const_index) % attList, 'coordinates', 'lonCell latCell')
      call mpas_add_att(r1aPtr(1) % attLists(1) % attList, 'standard_name', 'sea_surface_height_above_geoid')

```

Comment (xylar) [reply components/mpas-framework/src/tools/registry/gen_inc.c:1747]: Also, I think `const_index` needs to be `1` for a `var`.  See above and below.

**Review by xylar — CHANGES_REQUESTED** [review]
To test, I added:
```
                <var name="ssh" type="real" dimensions="nCells Time" units="m"
+                        standard_name="sea_surface_height_above_geoid"
+                        coordinates="lonCell latCell"
+                        bounds="lonVertex"
+                        cell_method="area: mean where sea"
+                        cell_measures="area: areaCell"
                         description="sea surface height"
                />
```
to the registry.

I'm seeing an error when I run on multiple processors pointing to this auto-generated line:
```
call mpas_add_att(r1aPtr(1) % attLists(const_index) % attList, 'bounds', 'lonVertex')
``` 
The error is:
```
Segmentation fault - invalid memory reference.
```
So not overly helpful.  I think the problem is `const_index` that should be `1` for a var.  See my comment in the appropriate code block.

File: components/mpas-framework/src/tools/registry/gen_inc.c [hunk 2]
```diff
@@ -1407,6 +1434,86 @@ int parse_var_array(FILE *fd, ezxml_t registry, ezxml_t superStruct, ezxml_t var
...
+				sprintf(temp_str, "%s", token);
+
+				while ( ( token = strsep(&string, "'") ) != NULL ) {
+					sprintf(temp_str, "%s''%s", temp_str, token);
+				}
+
+				free(tofree);
+
+				fortprintf(fd, "         call mpas_add_att(%s %% attLists(const_index) %% attList, 'bounds', '%s')\n", pointer_name_arr, temp_str);
+			}
```
Comment (xylar) [inline components/mpas-framework/src/tools/registry/gen_inc.c:1451]: I tried adding attributes to a `var_array` and they didn't come through in the code.  I added:
```xml
                <var_struct name="tracers" time_levs="2">
-                       <var_array name="activeTracers" dimensions="nVertLevels nCells Time" type="real" packages="activeTracersPKG" >
+                       <var_array name="activeTracers" dimensions="nVertLevels nCells Time" type="real" packages="activeTracersPKG"
+                                       coordinates="lonCell latCell"
+                                       bounds="lonVertex"
+                                       cell_method="area: mean where sea"
+                                       cell_measures="area: areaCell volume: volCell">
                                <var name="temperature" array_group="activeGRP" units="degrees Celsius"
+                                       standard_name="sea_water_potential_temperature"
                         description="potential temperature"
                                />
                                <var name="salinity" array_group="activeGRP" units="grams salt per kilogram seawater"
+                                       standard_name="sea_water_salinity"
                         description="salinity"
                                />
                        </var_array>

```
And the `standard_name` came through for each variable but the attributes on the `var_array` are absent in `src/inc/structs_and_variables.inc`.

**Comment by xylar** [comment]
I went ahead and fixed the `const_index` issues I indicated above in my local branch.  The next problem I run into is again `invalid memory referanc` on line:
https://github.com/E3SM-Project/E3SM/blob/2e2a00efcb9f2d121011012178f0b33f751a3602/components/mpas-framework/src/framework/mpas_pool_routines.F#L5048
This may relate to the other modifications that I think shouldn't be in this PR but that's as far as I got for now.

In debugging this, it might be helpful if I can give you some instructions on running a quick test via compass.  We can do that on Slack sometime but I don't have time right now.  If you have a quick E3SM test that would do the trick, feel free to use that.

**Comment by xylar** [comment]
@jonbob, have you had a chance to look into the issues I uncovered in my testing yet?

**Comment by jonbob** [comment]
@xylar -- my apologies, I have not gone back to rework this. I'll try to find time this week

File: components/mpas-framework/src/tools/registry/gen_inc.c [hunk 3]
```diff
@@ -608,26 +609,43 @@ int parse_namelist_records_from_registry(ezxml_t registry)/*{{{*/
...
 			nmloptposvals = ezxml_attr(nmlopt_xml, "possible_values");
 			nmloptindef = ezxml_attr(nmlopt_xml, "in_defaults");
 
 			if(strncmp(nmlopttype, "real", 1024) == 0){
 				fortprintf(fd, "      real (kind=RKIND) :: %s = %lf\n", nmloptname, (double)atof(nmloptval));
-				fortprintf(fcd, "      real (kind=RKIND) :: %s\n", nmloptname);
+				fortprintf(fcd, "      real (kind=RKIND), pointer :: %s\n", nmloptname);
+				if (strstr(nmloptname, "config_") == nmloptname ) {
+                   fortprintf(fcd, "      real (kind=RKIND) :: varinp_%s\n", nmloptname+7);
+                }
```
[the remark this answers is not in the document]
Comment (jonbob) [reply components/mpas-framework/src/tools/registry/gen_inc.c:626]: @xylar -- agreed. I don't know where they came from but I'm removing them.

**Comment by jonbob** [comment]
@xylar -- I've made all the fixes you suggested on a local branch and will test in the morning before pushing back. Thanks for your patience

**Comment by jonbob** [comment]
@xylar - I cleaned up following your suggestions and made sure the attributes are working, at least for vars, and pushed those changes to my branch. Now I'm trying to see what it will take for the vars to get attributes from the var_arrays....

**Comment by xylar** [comment]
Thanks @jonbob.  Keep me posted and let me know when you want me to review this again.

**Comment by jonbob** [comment]
@xylar - OK, I also have it working to have the vars under a var_array access the attributes when they are not directly assigned them. Or at least I've done this for everything except standard_name, mostly because I'm not sure that particular one makes sense? In any case, if I modify Registry for tracers as you did I see this in the output now:
````
	double temperature(Time, nCells, nVertLevels) ;
		temperature:long_name = "potential temperature" ;
		temperature:units = "degrees Celsius" ;
		temperature:bounds = "lonVertex" ;
		temperature:cell_measures = "area: areaCell volume: volCell" ;
		temperature:cell_method = "area: mean where sea" ;
		temperature:coordinates = "lonCell latCell" ;
		temperature:standard_name = "sea_water_potential_temperature" ;
	double salinity(Time, nCells, nVertLevels) ;
		salinity:long_name = "salinity" ;
		salinity:units = "grams salt per kilogram seawater" ;
		salinity:bounds = "lonVertex" ;
		salinity:cell_measures = "area: areaCell volume: volCell" ;
		salinity:cell_method = "area: mean where sea" ;
		salinity:coordinates = "lonCell latCell" ;
		salinity:standard_name = "sea_water_salinity" ;
````

**Comment by jonbob** [comment]
I included the capability to use the var_array standard_name attribute for corresponding vars as well, but only if the var standard_name attribute is NULL. This is the same approach as used for the other attributes, so var_array attributes are used only when the corresponding var attributes are missing. @xylar, I think this is ready for your review

**Comment by xylar** [comment]
Great @jonbob, I'll take a look at this as soon as I can.

**Review by xylar — APPROVED** [review]
@jonbob, thanks for working so hard on this!

I retested, this time on Chrysalis with all compilers and MPI (intel and impi, intel and openmpi, gnu and openmpi).  All results of the compass `pr` test suite are BFB with E3SM `master`.

I also reran my test in which I modified the `ssh` and active tracer variables in the registry to have several attributes.  
```xml
                <var name="ssh" type="real" dimensions="nCells Time" units="m"
+                        standard_name="sea_surface_height_above_geoid"
+                        coordinates="lonCell latCell"
+                        bounds="lonVertex"
+                        cell_method="area: mean where sea"
+                        cell_measures="area: areaCell"
                         description="sea surface height"
                />
```
and
```xml
                <var_struct name="tracers" time_levs="2">
-                       <var_array name="activeTracers" dimensions="nVertLevels nCells Time" type="real" packages="activeTracersPKG" >
+                       <var_array name="activeTracers" dimensions="nVertLevels nCells Time" type="real" packages="activeTracersPKG"
+                                       coordinates="lonCell latCell"
+                                       bounds="lonVertex"
+                                       cell_method="area: mean where sea"
+                                       cell_measures="area: areaCell volume: volCell">
                                <var name="temperature" array_group="activeGRP" units="degrees Celsius"
+                                       standard_name="sea_water_potential_temperature"
                         description="potential temperature"
                                />
                                <var name="salinity" array_group="activeGRP" units="grams salt per kilogram seawater"
+                                       standard_name="sea_water_salinity"
                         description="salinity"
                                />
                        </var_array>
```
Things look good:
```
$ ncdump -h /lcrc/group/e3sm/ac.xylar/compass_1.2/chrysalis/test_20220825/attributes/baroclinic_wit_ssh/ocean/baroclinic_channel/10km/default/forward/output.nc | more
...
	double ssh(Time, nCells) ;
		ssh:units = "m" ;
		ssh:bounds = "lonVertex" ;
		ssh:cell_measures = "area: areaCell" ;
		ssh:cell_method = "area: mean where sea" ;
		ssh:coordinates = "lonCell latCell" ;
		ssh:standard_name = "sea_surface_height_above_geoid" ;
		ssh:long_name = "sea surface height" ;
	double temperature(Time, nCells, nVertLevels) ;
		temperature:long_name = "potential temperature" ;
		temperature:units = "C" ;
		temperature:bounds = "lonVertex" ;
		temperature:cell_measures = "area: areaCell volume: volCell" ;
		temperature:cell_method = "area: mean where sea" ;
		temperature:coordinates = "lonCell latCell" ;
		temperature:standard_name = "sea_water_potential_temperature" ;
	double salinity(Time, nCells, nVertLevels) ;
		salinity:long_name = "salinity in grams salt per kilogram seawat
er" ;
		salinity:units = "1.e-3" ;
		salinity:bounds = "lonVertex" ;
		salinity:cell_measures = "area: areaCell volume: volCell" ;
		salinity:cell_method = "area: mean where sea" ;
		salinity:coordinates = "lonCell latCell" ;
		salinity:standard_name = "sea_water_salinity" ;
...
```

**Comment by xylar** [comment]
closed in favor of https://github.com/E3SM-Project/E3SM/pull/5183

**Pull request closed without merging**